content/blog/2021/10/02/MonitorTheSolver.adoc

```diff
@@ -0,0 +1,215 @@
+:jbake-type: post
+:jbake-tags: monitoring, production
+
+It's 11pm on Friday evening. Everything was working fine -- until now.
+Suddenly services are failing left and right, and your boss want to know why.
+One tool in achieving a diagnosis is to use a monitoring system. Let's add monitoring to OptaPlanner and see if we can diagnose this issue.
+
+== What is monitoring?
+
+_Monitoring_ is observing the quality of a service over time.
+Mointoring is similar to logging, except its output is more easily analyzed by a machine.
+The software being monitored output _metrics_, a numerical measurement of some aspect of the software.
+The metrics then get recorded in a monitoring system, where it can be graphed, trigger alerts, and be correlated with events.
+
+For instance, you can monitor a web service.
+Its metrics could include the number of hits to a particular endpoint, time to respond to a request, and the number of failures.
+If the number of hits suddenly increase, it could correspond to the Slashdot effect or a potential Denial of Service attack.
+If the average time to respond to a request is increasing because there are not enough servers to fulfill requests, you might want to acquire additional servers.
+Additionally, monitoring allow you to see the effect a change has on the entire system.
+If you notice the time to respond to a request with the same load as before decreases after a change was pushed, you can conclude the change improved performance.
+
+
+== Connecting monitoring systems to OptaPlanner
+
+OptaPlanner uses https://micrometer.io/[Micrometer] to collect its metrics.
+Micrometer then send the metrics to different monitoring systems using _registries_.
+Every monitoring system supported by Micrometer has its own registry.
+For instance, to connect OptaPlanner to https://prometheus.io/[Prometheus], you would
+add its registry as a dependency in the `pom.xml`:
+
+[source,xml]
+----
+<dependency>
+  <groupId>io.micrometer</groupId>
+  <artifactId>micrometer-registry-prometheus</artifactId>
+  <version>${micrometer.version}</version>
+</dependency>
+----
+
+Then you would add the Prometheus registry to the global registry:
+
+[source,java]
+----
+PrometheusMeterRegistry prometheusRegistry = new PrometheusMeterRegistry(PrometheusConfig.DEFAULT);
+
+try {
+   HttpServer server = HttpServer.create(new InetSocketAddress(8080), 0);
+   server.createContext("/prometheus", httpExchange -> {
+       String response = prometheusRegistry.scrape();
+       httpExchange.sendResponseHeaders(200, response.getBytes().length);
+       try (OutputStream os = httpExchange.getResponseBody()) {
+           os.write(response.getBytes());
+       }
+   });
+
+   new Thread(server::start).start();
+} catch (IOException e) {
+   throw new RuntimeException(e);
+}
+
+Metrics.addRegistry(prometheusRegistry);
+----
+
+If you use Quarkus or Spring Boot, this is already set up for you; you just need to add the registry dependency for your monitoring system.
```

Comment:
Can we start with a section for Quarkus - which will be very short I guess, but I would mention the actual artifactId to use (maybe show a pom.xml snippet of the dependency)

Then a section for Spring Boot and then this long section for plain Java?
Motivation: 1) most users that do monitoring will be using Quarkus or Spring Boot. 2) the plain java way is awfully long.

Reply:
I'd even argue that the "plain java" explanation isn't needed in the blog (only in our docs). I guesstimate that 99% of the users that actually do monitoring, will do it with the help of quarkus or spring boot.